src/http/client.cc

```diff
@@ -346,31 +346,57 @@ future<> client::make_request(request req, reply_handler handle, abort_source& a
     return do_make_request(std::move(req), std::move(handle), &as, std::move(expected));
 }
 
-future<> client::do_make_request(request req, reply_handler handle, abort_source* as, std::optional<reply::status_type> expected) {
-    return do_with(std::move(req), std::move(handle), [this, as, expected] (request& req, reply_handler& handle) mutable {
-        return with_connection([this, &req, &handle, as, expected] (connection& con) {
-            return do_make_request(con, req, handle, as, expected);
-        }, as).handle_exception_type([this, &req, &handle, as, expected] (const std::system_error& ex) {
-            if (as && as->abort_requested()) {
-                return make_exception_future<>(as->abort_requested_exception_ptr());
-            }
+static future<std::optional<uint32_t>> handle_request_exception(uint32_t& retries, abort_source* as, const std::system_error& ex) {
+    ++retries;
+    if (as && as->abort_requested()) {
+        return make_exception_future<std::optional<uint32_t>>(as->abort_requested_exception_ptr());
+    }
 
-            if (!_retry) {
-                return make_exception_future<>(ex);
-            }
+    auto code = ex.code().value();
+    if ((code != EPIPE) && (code != ECONNABORTED)) {
+        return make_exception_future<std::optional<uint32_t>>(ex);
+    }
+    return make_ready_future<std::optional<uint32_t>>(std::nullopt);
+}
 
-            auto code = ex.code().value();
-            if ((code != EPIPE) && (code != ECONNABORTED)) {
-                return make_exception_future<>(ex);
-            }
+future<> client::make_raw_request(request& req, reply_handler& handle, abort_source* as, std::optional<reply::status_type> expected) {
```

Comment:
If we need two make_request-s -- one that takes ownership and the other one that doesn't -- I'd suggest two overloads:

```c++
future<> make_request(request&& req, reply_handler&& handle, ...); // takes ownership
future<> make_request(request& req, reply_handler& handle, ...); // doesn't take ownership
```

Reply:
Will do in a separate PR